packages/devtools_app/lib/src/flutter/table.dart

```diff
@@ -446,36 +453,59 @@ class TreeTableState<T extends TreeNode<T>> extends State<TreeTable<T>>
     }
     assert(selectedNode == items[selectedNodeIndex]);
 
-    if (event.logicalKey == LogicalKeyboardKey.arrowDown ||
-        event.logicalKey == LogicalKeyboardKey.arrowUp) {
-      _moveSelection(event.logicalKey, scrollController, constraints.maxHeight);
+    if (event.logicalKey == LogicalKeyboardKey.arrowDown) {
+      _moveSelection(ScrollKind.down, scrollController, constraints.maxHeight);
+    } else if (event.logicalKey == LogicalKeyboardKey.arrowUp) {
+      _moveSelection(ScrollKind.up, scrollController, constraints.maxHeight);
+    }
+
+    // On left arrow collapse the row if it is expanded. If it is not, move
+    // selection to its parent.
+    else if (event.logicalKey == LogicalKeyboardKey.arrowLeft) {
+      if (selectedNode.isExpanded) {
+        _toggleNode(selectedNode);
+      } else {
+        _moveSelection(
+            ScrollKind.parent, scrollController, constraints.maxHeight);
+      }
+    }
+
+    // On right arrow expand the row if possible, otherwise move selection down.
+    else if (event.logicalKey == LogicalKeyboardKey.arrowRight) {
+      if (!selectedNode.isExpanded) {
+        _toggleNode(selectedNode);
+      } else {
+        _moveSelection(
+            ScrollKind.down, scrollController, constraints.maxHeight);
+      }
     }
-    // TODO(gmoothart): Add expand/collapse with left/right keys
   }
 
   void _moveSelection(
-    LogicalKeyboardKey key,
+    ScrollKind scrollKind,
     ScrollController scrollController,
     double viewportHeight,
   ) {
     // get the index of the first item fully visible in the viewport
     final firstItemIndex =
         (scrollController.offset / _Table.defaultRowHeight).ceil();
 
+    // '-1' in the following calculations is needed because the header row
+    // occupies space in the viewport so we must subtract it out.
     final minCompleteItemsInView =
         (viewportHeight / _Table.defaultRowHeight).floor() - 1;
     final lastItemIndex = firstItemIndex + minCompleteItemsInView - 1;
     int newSelectedNodeIndex;
 
-    if (key == LogicalKeyboardKey.arrowDown) {
-      // move selection down if there is a node lower down.
+    if (scrollKind == ScrollKind.down) {
```

Comment:
could use a switch statement here as well to make it clear that there are three possible cases

Reply:
Done.